Add localized LED flash effect on keypress to Massdrop ALT

reywood opened the pull request:
Hi there. This pull request evolved from trying to implement some custom LED effects on the Massdrop ALT. The code here causes a flash to ripple out from a key's LED when pressed. I had to modify the arm_atsam code to make this possible, though I tried to make as few changes to it as possible, only enough to make more diverse types of effects possible. As such, I'd be interested in hearing from @patrickmt on these changes. [Here's a video](https://imgur.com/Ea6av2j) of the effect.

drashna commented:
Would it be possible to rename the functions to match `https://docs.qmk.fm/#/feature_rgb_matrix?id=custom-layer-effects` type naming.

This is mostly as a precursor to start migrating the code to the QMK RGB Matrix feature, so we don't have two matrix functions.

patrickmt commented on tmk_core/protocol/arm_atsam/led_matrix.h:
```diff
@@ -86,6 +86,7 @@ typedef struct led_disp_s {
 } led_disp_t;
 
 uint8_t led_matrix_init(void);
+void led_matix_init_user(void);
```
Misspelled "matrix" in function name

drashna replied:
```suggestion
void led_matrix_init_user(void);
```

patrickmt commented on keyboards/massdrop/alt/keymaps/reywood/led_matrix_user.c:
```diff
@@ -0,0 +1,193 @@
+static uint8_t map_key_position_to_led_index(uint8_t col, uint8_t row);
+
+
+void led_matix_init_user(void) {
```
Misspelled "matrix" in function name

drashna replied:
```suggestion
void led_matrix_init_user(void) {
```

patrickmt commented on tmk_core/protocol/arm_atsam/led_matrix.c:
```diff
@@ -459,13 +461,19 @@ uint8_t led_matrix_init(void)
     //Run led matrix code once for initial LED coloring
     led_cur = 0;
-    led_matrix_run((led_setup_t*)led_setups[led_animation_id]);
+    led_matix_init_user();
```
Misspelled "matrix" in function name

drashna replied:
```suggestion
    led_matrix_init_user();
```

patrickmt commented on tmk_core/protocol/arm_atsam/led_matrix.c:
```diff
@@ -459,13 +461,19 @@ uint8_t led_matrix_init(void)
 }
 
+__attribute__ ((weak))
+void led_matix_init_user(void) {
```
Misspelled "matrix" in function name

drashna replied:
```suggestion
void led_matrix_init_user(void) {
```

patrickmt commented:
Looks nice! I have your keymap running on an ALT here. I made a few comments on "matrix" being misspelled in user init function. Also the default keymap works as expected for ALT and CTRL keyboards. I checked the changes on Massdrop's fork for the LED code and looks like there would be no conflicts there either.

reywood commented:
@patrickmt Good catch on that typo!

@drashna Considering your comment about renaming, should I rename `led_mat[r]ix_init_user` to something like `rgb_matrix_init_user`? Is there anything else in particular you want me to rename?

drashna commented:
I would say that `rgb_matrix_init_user` would be the better name, I think.  Any input on that @patrickmt?

patrickmt commented:
Getting it more aligned with QMK is good, so `rgb_matrix_init_user` is fine by me!

reywood commented:
Alright, I updated the function name to `rgb_matrix_init_user` and updated my file names to match. I didn't update the other functions that were already present in the `arm_atsam` code. I could update those as well, but don't want to step on any work that @patrickmt may already have in progress.

drashna commented:
@reywood would you mind making a PR with the same changes to the CTRL board too?

reywood commented:
I'll take a look. I don't have a CTRL to test on though, so no guarantees that the code will work 😬

patrickmt commented:
I can take care of the testing if you do take this on.

valen214 commented:
@reywood I have been working with similar effect recently. It is kind of funny because I am doing this with CTRL.
My implementation only requires to modify `[keymap.c](https://github.com/valen214/Massdrop-Ctrl-Keyboard-Ripple-LED/blob/master/keyboards/massdrop/ctrl/keymaps/custom/keymap.c)` and nothing else.

I think I might take some time to have a look on your code if I can get something new.
BTW, would you mind to help me testing out my code with ALT keyboard? (just copy the `keymap.c` file to `./keyboards/massdrop/alt/keymaps/<profile name>` and build is good enough, I hope)
